### PR TITLE
Web Support and Refactoring

### PR DESCRIPTION
**Pull request by personalizedrefrigerator**
## Overview
This pull request adds support for web-based favicons and [Progressive Web App](https://developers.google.com/web/fundamentals/web-app-manifest/#icons) launcher icons.

## Summary of Changes
* Added `web.dart`. This script crops, resizes, and copies icons specified in configuration files.
* Added tests for `web.dart`. Additional tests were necessary for the new configuration script.
* Updated example projects.
* Refactoring & general bug fixes.

## Notes
 * The page's updated favicon is not shown when running `flutter run -d chrome` or `flutter run -d web-server`. The project must first be built for web, then served from the build directory (see [the README](https://github.com/personalizedrefrigerator/flutter_launcher_icons#launcher-icons-not-present-in-debug-mode)).
 * Several directories that are automatically generated by `flutter create .` have been removed from the example projects and added to the project's `.gitignore` file. This minimizes the number of changes committed by PRs that modify the example projects, but does mean users will have to run `flutter create .` in each of the example project directories (see the README in the default example project).
     - Most of the `264` files GitHub lists as "changed" are deleted files! Most were in the example project's `ios` and `android` directories! Please, don't let this large number stop you from reviewing this PR!

### CODE REVIEW

**Comment by personalizedrefrigerator** [comment]
## Supporting multiple icon sets
The Android and iOS icon-distribution scripts both support sets of different launcher icons, aside from just the default. The web-favicon and [PWA](https://developer.mozilla.org/en-US/docs/Web/Manifest) icon-generation script does not support this.

## Ways support might be added
 * Creation of a backup folder to where current icons are copied before replacement of the default icon set. This could permit saving `favicon.png` in addition to the files in the `icons` folder.
 * Overwrite `favicon.png` without saving it, but change the launcher icon folder to the new, requested folder assuming that `favicon.png` can be generated from the contents of the `icons` folder.

File: example/.gitignore
Comment (bean5) [inline example/.gitignore]: Adding this file seems to fall outside scope of ticket.

**Comment by bean5** [comment]
@personalizedrefrigerator This looks good to me, but I'm not approved to approve and merge. Also, I'm wondering why iOS and Android can handle my icon, but web cannot. Maybe I've formatted my main icon incorrectly?

**Comment by personalizedrefrigerator** [comment]
@bean5 Thanks for letting me know! I'll work on that. What does your `pubspec.yaml` look like?

**Comment by bean5** [comment]
Looks good to me! Web build fine this time!

**Comment by bean5** [comment]
@MarkOSullivan94 was there something you wanted changed before merging this?

**Comment by MarkOSullivan94** [comment]
I actually haven't got round to testing out Flutter Web yet and feel merging this PR would require someone from the community to step up and keep an eye on any Web related issues / suggestions

Happy to merge in either @bean5 or @personalizedrefrigerator be happy to do this?

**Comment by personalizedrefrigerator** [comment]
@MarkOSullivan94 I would be happy to help!

**Comment by bean5** [comment]
@personalizedrefrigerator 

`7.5.0` would make more sense now that I know that the web cropping issue wasn't indicative of a compatibility issue.

**Comment by bean5** [comment]
@MarkOSullivan94 whenever I start make a new project, I'm going to be generating an icon for web, so testing will happen naturally for web.

While for most projects I believe that a `web` branch/fork should exist until flutter supports web in their `stable` channel, I don't think that is necessary for this project because it is a dev dependency. I don't see any harm in merging this once the version number is revised to reflect that it is backwards-compatible (ie `7.*`).

**Comment by bean5** [comment]
@MarkOSullivan94 feel free to add me as a contributor if it helps. I can't commit to an official amount of time, but adding me shouldn't hurt anything. :)

**Comment by personalizedrefrigerator** [comment]
Whoops! My fault! The version code was `0.8.0` and `0.7.5`, not `8.0.0` and `7.5.0`! I support the switch to `0.8.0` but understand any wish to change this.

**Comment by bean5** [comment]
Oh, `0.8.0` is fine. It shows backwards-compatibility.

LGTM

**Comment by bean5** [comment]
Until this gets merged in, I'm relegated to using the fork instead. Just sayin'.

**Comment by TheJulianJES** [comment]
Love it, but I did get the following error when I tried it:
```
Overwriting web favicon and launcher icons...
FileSystemException: Cannot open file, path = 'web/icons/Icon-192.png' (OS Error: No such file or directory, errno = 2)
```
Creating the icons folder in the web directory fixed it though!

**Comment by cocacrave** [comment]
Does the web icon come out really pixelated for anyone else?
Comparing same size for android/ios vs web generated icons, they come out pixelated.
I'm using Flutter beta v1.15.17 on MacOS 10.15.3.

```
dev_dependencies:
  flutter_test:
    sdk: flutter
  flutter_launcher_icons: ^0.7.4

dependency_overrides:
  flutter_launcher_icons:
    git:
      url: https://github.com/personalizedrefrigerator/flutter_launcher_icons
      ref: master
```

Am I doing something wrong?

**Comment by personalizedrefrigerator** [comment]
Thanks! I'm looking into it!

**Comment by cocacrave** [comment]
@personalizedrefrigerator Thanks, here is the rest of pubspec.yaml for flutter_icons:
```
flutter_icons:
  web: true
  ios: true
  android: true
  image_path_web: "assets/icons/launcher_icon_web.png"
  image_path_ios: "assets/icons/launcher_icon_ios.png"
  image_path_android: "assets/icons/launcher_icon_android.png"
  adaptive_icon_background: "assets/icons/adaptive_icon_background.png"
  adaptive_icon_foreground: "assets/icons/adaptive_icon_foreground.png"
```

The launcher icons for ios and android looks fine but the generated `web/favicon.png` (16x16) and the two icons in `web/icons/` named `Icon-192.png` and `Icon-512.png` looks pixelated. I just replaced those 3 icons manually for now.

Thanks for looking into it.

**Comment by personalizedrefrigerator** [comment]
I'm moving this discussion to an [issue in the forked repository](https://github.com/personalizedrefrigerator/flutter_launcher_icons/issues/1). I'm having difficulties reproducing the issue!

File: example/android/app/src/debug/AndroidManifest.xml
Comment (MarkOSullivan94) [inline example/android/app/src/debug/AndroidManifest.xml:7]: Nice addition using this only in the debug `AndroidManifest.xml`

File: example/android/app/src/main/AndroidManifest.xml
Comment (MarkOSullivan94) [inline example/android/app/src/main/AndroidManifest.xml:10]: I always end up with way too many "example" named apps on my emulators 😂 

```suggestion
        android:label="flutter_launcher_icons"
```

File: example/android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java
Comment (MarkOSullivan94) [inline example/android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java:4]: Are these used?

File: example/android/build.gradle
Comment (MarkOSullivan94) [inline example/android/build.gradle:2]: Can update this to latest, not sure what it is but I know at least `1.3.70` has been released

File: example/ios/Flutter/flutter_export_environment.sh
Comment (MarkOSullivan94) [inline example/ios/Flutter/flutter_export_environment.sh:2]: My bad for putting this into version control

File: example/ios/Runner/Info.plist
Comment (MarkOSullivan94) [inline example/ios/Runner/Info.plist:14]: Again tiny tweak but helps me prevent having too many example apps 😅 

```suggestion
	<string>flutter_launcher_icons</string>
```

File: example/android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java
Comment (personalizedrefrigerator) [reply example/android/app/src/main/java/io/flutter/plugins/GeneratedPluginRegistrant.java:4]: The `Keep` and `NonNull` annotations were auto-generated -- I was having trouble getting the older example application to work, so re-created it using `flutter create`. These annotations appear to be used on lines `12` and `14`.

File: example/android/build.gradle
Comment (personalizedrefrigerator) [reply example/android/build.gradle:2]: Okay! Thanks!

File: example/android/app/src/debug/AndroidManifest.xml
Comment (personalizedrefrigerator) [reply example/android/app/src/debug/AndroidManifest.xml:7]: Thanks, but give credit to the Flutter team! This change came from re-creating the example application using `flutter create`.

File: example/android/build.gradle
Comment (personalizedrefrigerator) [reply example/android/build.gradle]: Resolved in a later commit.

**Comment by bean5** [comment]
Glad to see thing smoving!

**Comment by llucax** [comment]
Hi, it seems like this PR is stalled. Is there a particular reason for that, can something be done to get it merged? There are already 3 issues requesting this feature: #209, #183, #126 :)

**Comment by personalizedrefrigerator** [comment]
Hi,
    I think this hasn't been merged in because I regenerated the example project (I was having trouble getting it to work). 
    Unfortunately, the example project contains a lot of code that would need to be reviewed before a merge. 
    I can try replacing the example project with that in the current `flutter_launcher_icons`. 
Thank you for bringing this up! I haven't been using Flutter recently and so didn't prioritize this pull request!

**Comment by bean5** [comment]
@personalizedrefrigerator Whatever it takes, it would be nice to have this finally merged in.

**Comment by personalizedrefrigerator** [comment]
@bean5 Not yet -- a cropping issue has been reported. I should fix that first. :)

File: example/ios/Flutter/flutter_export_environment.sh
Comment (personalizedrefrigerator) [reply example/ios/Flutter/flutter_export_environment.sh]: This should be fixed!

**Comment by personalizedrefrigerator** [comment]
It looks like non-square images are squished, rather than padded. This also happens when iOS launcher icons are created using the version of `flutter_launcher_icons` in master.

### Original (iOS)

image

### Generated (iOS)

image

Is this expected behavior?

**Comment by personalizedrefrigerator** [comment]
@MarkOSullivan94 The web branch is somewhat behind the main branch. As the fork has recently pulled from the main branch, it should be easier to review changes relative to master. Feel free to change the base back to web.

**Comment by personalizedrefrigerator** [comment]
@slightfoot @MarkOSullivan94 It would be really nice to have web support in `flutter_launcher_icons`. Could you please review this PR?

**Review by MarkOSullivan94 — CHANGES_REQUESTED** [review]
Hi @personalizedrefrigerator, appreciate your patience!

Hopefully not too many merge conflicts since when you started this PR 🤞 

I've tested it out and it's working great! Awesome job with this 😃 

Some feedback: 

**Include example web project**

I know you stated that you excluded the `flutter create .` step in each of the example directions, the example projects are there so anyone interested in using Flutter Launcher Icons or contribute to the project can quickly run the examples.

I'd recommend adding the default Flutter example project into another directory called `web` and ensure it's ready for anyone to quickly run to test out the web support.

**Include check to determine if Flutter Web is already enabled**

I appreciate you adding the info in the Troubleshooting section however there might be some people trying web for the first time, best to include that they can check `flutter devices` to see if web support has been enabled and if not, provide a link to the official docs on how to enable support for it in their project

File: lib/web.dart
Comment (personalizedrefrigerator) [reply lib/web.dart:20]: This comment might not be necessary! Other platforms need square icons, too, and the names of the output icons suggest that they will be square. I have removed it.

File: example/flavors/pubspec.yaml
Comment (personalizedrefrigerator) [reply example/flavors/pubspec.yaml:21]: Okay! I can have a copy here and a copy in the Troubleshooting section!

Comment (personalizedrefrigerator) [reply example/flavors/pubspec.yaml:21]: Done!

**Comment by personalizedrefrigerator** [comment]
@MarkOSullivan94 Thank you for the feedback!

I think I've made most of the changes you requested, but would like to clarify your suggestion that I include an example web project.

 * Are you requesting that I run `flutter create default_example/; flutter create flavors/` and version control the result?
    - I personally think that omitting directories that can be generated makes it easier to contribute and to find relevant parts of the example project, but this is just personal preference. 😄
    - If version controlling the results of `flutter create .`, after finalizing the contents of the example project (things not generated by `flutter create`), I would prefer a maintainer run `flutter create` to generate the contents of the example projects. (I'm paranoid & having maintainers create generated code -- in general -- makes me less worried about malicious changes to generated code 😄).
 * The both example projects (at present) demonstrate web support. Are you requesting that I remove `web: true`/`web: false` from these and instead have a separate project for `web` support?

**Comment by bean5** [comment]
Glad to see this moving again!

**Comment by MarkOSullivan94** [comment]
> The both example projects (at present) demonstrate web support. Are you requesting that I remove web: true/web: false from these and instead have a separate example project for web support?

This! 😄 

I agree there's no point committing all the extra files for the original examples because someone might only want to test out Flutter Launcher Icons for Android for example.

Plus adding extra flags in the default examples yaml file will probably end up making everyone think that you need the `web: false` flag all the time when it's fine to omit it.

An isolated example for web which is ready to for anyone to run immediately would be better.

**Comment by personalizedrefrigerator** [comment]
@MarkOSullivan94 That should be done!

**Comment by personalizedrefrigerator** [comment]
@MarkOSullivan94 I think I've made the changes you requested! Is there anything else you would like me to change?

Again, thank you for taking the time to review this PR!

**Comment by personalizedrefrigerator** [comment]
@MarkOSullivan94 If the number of deleted files is making this difficult to review, I can restore them. You could then re-delete the files after reviewing this PR.

I'm not sure if this is an issue, but it would be very nice to have this merged!

**Comment by MarkOSullivan94** [comment]
@personalizedrefrigerator apologies super busy atm! Should get a chance to review this coming weekend 🤞

**Comment by personalizedrefrigerator** [comment]
That's okay! I understand!

**Comment by personalizedrefrigerator** [comment]
@MarkOSullivan94 To limit the size of the diff, I'm restoring the old `default` and `flavors` examples. The new example projects are called `default_example` (`flutter create .` didn't work when called `default` for this one) and `flavors_renameme`. 

After reviewing, please remove the `examples/default` directory and replace `examples/flavors` with `examples/flavors_renameme`. Thank you!

File: example/web/lib/main.dart
Comment (personalizedrefrigerator) [inline example/web/lib/main.dart:1]: This file is just the default `main.dart` generated by `flutter create .`. While `flutter create .` **is** in the instructions, it might still be useful to have this committed.

File: lib/android.dart
Comment (personalizedrefrigerator) [inline lib/android.dart:31]: `createDefaultIcons` and `createAdaptiveIcons` were moved! See below.

File: lib/icon_template.dart
Comment (personalizedrefrigerator) [inline lib/icon_template.dart:31]: This replaces the different `IconTemplate` classes in `android.dart` and `ios.dart`. These classes were small, but having a central `IconTemplate` class will (I hope!) help with code re-use.

File: lib/ios.dart
Comment (personalizedrefrigerator) [inline lib/ios.dart:33]: Just like in `android.dart`, `createIcons` overrides a method in `AbstractPlatform`. The following code has been moved.

File: lib/main.dart
Comment (personalizedrefrigerator) [inline lib/main.dart:30]: I think these changes are from `dartfmt`...

File: lib/main.dart
Comment (personalizedrefrigerator) [inline lib/main.dart:64]: This should address #215...

File: lib/main.dart
Comment (personalizedrefrigerator) [inline lib/main.dart:122]: There was an issue where tests would sometimes fail when running in parallel. 
One of the tests in the test suite changed the current working directory for **all tests**. When other tests required the current working directory to be something specific ran, this could cause them to fail.

My solution was to not actually change the current working directory. It does make the code somewhat more confusing though...

File: pubspec.yaml
Comment (personalizedrefrigerator) [inline pubspec.yaml:17]: It might be worthwhile to add a `CONTRIBUTING.md` that explains how to run the tests.

I was initially unaware that I had to run `flutter pub test` rather than `flutter test` to test this package.

File: test/android_test.dart
Comment (personalizedrefrigerator) [inline test/android_test.dart:173]: All of the changes in `android_test.dart` should be from `dartfmt`.

File: bin/main.dart
Comment (personalizedrefrigerator) [inline bin/main.dart:5]: This and some additional tests added to the test suite should address #207...

**Comment by personalizedrefrigerator** [comment]
To put some important notes at the very bottom of this page,

### After reviewing, please do the following:

 * **Remove the `examples/default` directory**
 * **Replace `examples/flavors` with `examples/flavors_renameme`.**

Thank you!

File: README.md
Comment (personalizedrefrigerator) [inline README.md:12]: Make sure to update the release date (and possibly select a different version, if another is more appropriate).

File: CHANGELOG.md
Comment (personalizedrefrigerator) [inline CHANGELOG.md:3]: I'm not sure what an appropriate version number would be... Consider `0.9.0` a placeholder for whatever seems appropriate.

File: README.md
Comment (personalizedrefrigerator) [inline README.md:247]: If `flavors_renameme` is renamed to something other than `flavors`, this link will need to be updated!

**Comment by personalizedrefrigerator** [comment]
To put some important notes below my above comments,

### After reviewing, please do the following:

 * **Remove the `examples/default` directory**
 * **Replace `examples/flavors` with `examples/flavors_renameme`.**

Thank you!

**Comment by MarkOSullivan94** [comment]
> The new example projects are called default_example (flutter create . didn't work when called default for this one) and flavors_renameme

Wait, why have you needed to recreate the default examples?

File: CHANGELOG.md
Comment (MarkOSullivan94) [reply CHANGELOG.md]: 0.9.0 will be the next release number with web support, I consider this to be big new functionality but still missing a couple features for me to be happy to have Flutter Launcher Icons bumped to v1.0.0

File: README.md
Comment (MarkOSullivan94) [reply README.md]: I always update the release date before release 😃

File: bin/main.dart
Comment (MarkOSullivan94) [reply bin/main.dart:5]: Thanks for the fix! 🎉

File: test/android_test.dart
Comment (MarkOSullivan94) [reply test/android_test.dart:173]: Thanks for the note!

**Review by MarkOSullivan94 — COMMENTED** [review]
Left a couple of comments for you to check out!

Got a feeling it wont be long until this is merged 🙌

**Comment by personalizedrefrigerator** [comment]
> > The new example projects are called default_example (flutter create . didn't work when called default for this one) and flavors_renameme
> 
> Wait, why have you needed to recreate the examples which were there before this PR?

GitHub's diff was showing every deleted file -- I added copies of the old example projects to limit the size of the diff GitHub displays.

File: lib/constants.dart
Comment (personalizedrefrigerator) [reply lib/constants.dart]: Done! The output looks somewhat strange with the leading newline, though:

```
Flavor: development
• Creating default icons Android
• Overwriting the default Android launcher icon with a new icon
• Building iOS launcher icon for development

Flavor: integration
• Creating default icons Android
• Overwriting the default Android launcher icon with a new icon
• Building iOS launcher icon for integration

Flavor: production

⚠ WARNING:  Not generating icons for web. Platform web is in config, but there is no matching directory. If you want to generate icons for web, make sure you have support for web enabled and have run `flutter create .` in the root directory of your project.
• Creating default icons Android
• Overwriting the default Android launcher icon with a new icon
• Building iOS launcher icon for production

✓ Successfully generated launcher icons for flavors
```

As such, I have added a trailing newline to generateWarning.

File: example/web/lib/main.dart
Comment (personalizedrefrigerator) [reply example/web/lib/main.dart]: I've removed these!

File: lib/main.dart
Comment (personalizedrefrigerator) [reply lib/main.dart:132]: Done!

File: example/web/pubspec.yaml
Comment (personalizedrefrigerator) [reply example/web/pubspec.yaml:25]: I'm confused... I don't think there is a configuration that updates **just** the favicon...
Please clarify.

Additionally, looking at the generated `index.html` file, 

```html
  <!-- Favicon -->
  <link rel="icon" type="image/png" href="favicon.png"/>
```

is present.

As such, I think I can implement this!!!

File: lib/constants.dart
Comment (MarkOSullivan94) [reply lib/constants.dart]: Yeah looks weird with the leading newline, trailing newline to `generateWarning` looks better! 😄

**Comment by MarkOSullivan94** [comment]
> GitHub's diff was showing every deleted file — I added copies of the old example projects to limit the size of the diff GitHub displays.

Size of the diff is fine don't worry 😃

File: lib/web.dart
Comment (MarkOSullivan94) [inline lib/web.dart:34]: Not necessarily replacing the launcher icon if they have only setup the config file to replace the web favicon

```suggestion
      printStatus('Overwriting web favicon');
```

File: example/web/pubspec.yaml
Comment (MarkOSullivan94) [reply example/web/pubspec.yaml:25]: > I'm confused... I don't think there is a configuration that updates just the favicon...
Please clarify.

This config would only update the favicon 🙂 

```
flutter_icons:
  image_path_web: "assets/images/icon-710x599-web.png"
  web: true # Can't specify file name here at present.
```

**Comment by MarkOSullivan94** [comment]
Last change I'd recommend: Probably best to go ahead and execute the `flutter create .` command in the web example 

This would mean one less step to run the web example 🙂

File: example/web/pubspec.yaml
Comment (personalizedrefrigerator) [reply example/web/pubspec.yaml:25]: Oh! That updates web launcher icons as well! (E.g. for installing a website as a progressive web app).

**Comment by personalizedrefrigerator** [comment]
Done! 

78 commits is **a lot**. Would you like me to try to merge/fixup groups of commits?

**Comment by personalizedrefrigerator** [comment]
At present, unlike the Android and iOS launcher icon generators, `web.dart` cannot be given an alternate icon path (e.g. `web: favicon_test.png`). I think this is better left as the topic of a followup PR.

Would you like me to try to add this functionality?

**Comment by MarkOSullivan94** [comment]
> Done!
> 
> 78 commits is **a lot**. Would you like me to try to clean up the commit history (i.e. using `git reflog -i` or similar)?

Yep that would be appreciated!

> At present, unlike the Android and iOS launcher icon generators, web.dart cannot be given an alternate icon path (e.g. 
> web: favicon_test.png). I think this is better left as the topic of a followup PR.
>
> Would you like me to try to add this functionality?

I agree this is better to leave for a follow up PR

File: example/web/pubspec.yaml
Comment (MarkOSullivan94) [inline example/web/pubspec.yaml:25]:
```suggestion
  image_path_web: "assets/images/icon-710x599-web.png"
  web: true # Can't specify file name here at present.
```

Comment (MarkOSullivan94) [reply example/web/pubspec.yaml:25]: Of course 😅 My bad!

Comment (personalizedrefrigerator) [reply example/web/pubspec.yaml:25]: Okay! I can also get rid of the `android` and `ios` directories!

**Comment by MarkOSullivan94** [comment]
@personalizedrefrigerator heads up your commits removes some important files for the default example - guessing you done this before adjusting the `.gitignore`?

**Comment by personalizedrefrigerator** [comment]
> @personalizedrefrigerator heads up your commits removes some important files for the default example - guessing you done this before adjusting the `.gitignore`?

I think the default example only needs `pubspec.yaml`, `assets/images/` and `README.md` -- everything else is created by `flutter create .`.

**Comment by MarkOSullivan94** [comment]
> everything else is created by `flutter create .`.

Less steps needed for anyone to run the examples the better 🙂

**Comment by qbait** [comment]
Hey guys, do you have ETA for that by any chance?

**Comment by bean5** [comment]
Flutter 2.0 was announced today. It officially supports web, Linux, and other platforms. At pub.dev this project promises to be compatible for web at but I don't think it is--unless this is a duplicate branch and it was already done.

I just checked. The default is not building for web. 

How can I help get this done? What is blocking this from being finished up?

**Comment by leithhobson** [comment]
> Flutter 2.0 was announced today. It officially supports web, Linux, and other platforms. At pub.dev this project promises to be compatible for web at but I don't think it is--unless this is a duplicate branch and it was already done.
> 
> I just checked. The default is not building for web.
> 
> How can I help get this done? What is blocking this from being finished up?

The package is stated as compatible as your web project will not be broken by it (in my understanding) but yes, it is certainly time that the web support feature is implemented.

**Comment by Jwiggiff** [comment]
Is there any update/progress on this? Is there anything I could do to help? Would really love some web support :)

**Comment by personalizedrefrigerator** [comment]
I haven't been using Flutter recently! I'll try to resolve conflicts for this PR (but might have to close it & open another -- resolving conflicts will probably involve a rebase and users of the fork may be having Flutter fetch a specific commit).

**Comment by alcemirsantos** [comment]
Just got here coming from Issue #126. I think it would be interesting to link the issue to this PR since they are definitely related and there are other user asking about this there.

**Comment by personalizedrefrigerator** [comment]
Closing in favor of #189

**Pull request closed without merging**
